gitpod-io/gitpod
# Limit resources based on workspace class

## Description
This PR adds support for cpu limiting based on the workspace class of the workspace. The configuration for each workspace class contains a cpu limiting section which describes the minimal resources the workspaces gets and the burst limit. These values will be set as annotations on the workspace pod. ws-daemon will then use these annotations to distribute cpu resources

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/10981

## How to test
- Open workspace
- Check that the workspace pod contains two annotations: `gitpod.io/cpuMinLimit` and `gitpod.io/cpuBurstLimit`
- SSH into the harvester vm and find the cgroup for the workspace pod
- Check that the min limit has been applied
- Install stress-ng in the workspace
- Run `stress-ng --cpu 4`
- After a few seconds you should see that the burst limit has been applied to the cgroup of the workspace
- Use kubectl annotate to change the annotations
- The cgroup of the workspace should have been updated with the new values
- Deactivate feature flag in configcat
- Start workspace
- Check that no limit annotations are present
- Find the cgroup of the workspace
- Check that the limits configured in the ws-daemon configmap are applied

## Release Notes

```release-note
None
```

## Note
This relies on on a change in the werft files, that is why the main build failed. The custom build with the updated werft files (below) is successful however.

## Werft options:
